pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -0,0 +1,394 @@
+						continue
+					}
+				}
+				con.Routes = routes
+				if adsDebug {
+					log.Infof("ADS:RDS: REQ %s %s routes: %d", peerAddr, con.ConID, len(con.Routes))
+				}
+				err := s.pushRoute(con)
+				if err != nil {
+					return err
+				}
+
+			case EndpointType:
+				clusters := discReq.GetResourceNames()
+				if len(clusters) == len(con.Clusters) || len(clusters) == 0 {
+					if discReq.ErrorDetail != nil {
+						log.Warnf("ADS:EDS: ACK ERROR %v %s %v", peerAddr, con.ConID, discReq.String())
+					}
+					if edsDebug {
+						// Not logging full request, can be very long.
+						log.Infof("ADS:EDS: ACK %s %s %s %s", peerAddr, con.ConID, discReq.VersionInfo, discReq.ResponseNonce)
+					}
+					if len(con.Clusters) > 0 {
+						// Already got a list of clusters to watch and has same length as the request, this is an ack
+						continue
+					}
+				}
+				con.Clusters = clusters
+				for _, c := range con.Clusters {
+					s.addEdsCon(c, con.ConID, con)
+				}
+				if adsDebug {
+					log.Infof("ADS:EDS: REQ %s %s clusters: %d", peerAddr, con.ConID, len(con.Clusters))
+				}
+				err := s.pushEds(con)
+				if err != nil {
+					return err
+				}
+
+			default:
+				log.Warnf("ADS: Unknown watched resources %s", discReq.String())
+			}
+
+			if !con.added {
+				con.added = true
+				s.addCon(con.ConID, con)
+				defer s.removeCon(con.ConID, con)
+			}
+		case <-con.pushChannel:
+			// It is called when config changes.
+			// This is not optimized yet - we should detect what changed based on event and only
+			// push resources that need to be pushed.
+			if con.CDSWatch {
+				err := s.pushCds(*con.modelNode, con)
+				if err != nil {
+					return err
+				}
+			}
+			if len(con.Routes) > 0 {
+				err := s.pushRoute(con)
+				if err != nil {
+					return err
+				}
+			}
+			if len(con.Clusters) > 0 {
+				err := s.pushEds(con)
+				if err != nil {
+					return err
+				}
+			}
+			if con.LDSWatch {
+				err := s.pushLds(*con.modelNode, con)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+}
+
+// adsPushAll implements old style invalidation, generated when any rule or endpoint changes.
+// Primary code path is from v1 discoveryService.clearCache(), which is added as a handler
+// to the model ConfigStorageCache and Controller.
+func adsPushAll() {
+	// First update all cluster load assignments. This is computed for each cluster once per config change
+	// instead of once per endpoint.
+	edsClusterMutex.Lock()
+	// Create a temp map to avoid locking the add/remove
+	cMap := map[string]*EdsCluster{}
+	for k, v := range edsClusters {
+		cMap[k] = v
+	}
+	edsClusterMutex.Unlock()
+
+	for clusterName, edsCluster := range cMap {
+		updateCluster(clusterName, edsCluster)
+	}
+
+	// Push config changes, iterating over connected envoys. This cover ADS and EDS(0.7), both share
+	// the same connection table
+	adsClientsMutex.RLock()
+	// Create a temp map to avoid locking the add/remove
+	tmpMap := map[string]*XdsConnection{}
```

Comment:
size the map correctly to avoid re-allocs while holding the lock: `tmpMap := make(map[string]*XdsConnection, len(adsClients))`

Reply:
done